File: examples/test_lidar/lidar.py
import serial
import mraa

class Lidar(object):
  'class to represent vx-11 lidar'

  distances = [0] * 360
  packet = [0] * 22

  startReading = False
  packetNumber = 0
  speed = 0
  packetIndex = 0
  distanceIndex = 0

  def __init__(self):
    self.uart = mraa.Uart(0)
    self.ser = serial.Serial(self.uart.getDevicePath(), 115200)

  def read(self):
    waiting = self.ser.inWaiting()
    if waiting > 0:
      data = self.ser.read(waiting)

      for byte in data:
        byte = int(ord(byte))

        if self.startReading:
          if self.packetIndex == 21:
            self.startReading = False
            self.processEndOfPacket()
            # let the calling function know we're done
            return self.distanceIndex >= 359
          else:
            self.packet[self.packetIndex] = byte
            self.packetIndex += 1
            return False

        if not self.startReading and byte == 0xFA:
          self.startReading = True
          self.packetIndex = 0
          self.packet[self.packetIndex] = byte
          self.packetIndex += 1
          return False
        else:
          print "fuck"

  def processEndOfPacket(self):
    newPacketNumber = self.packet[1]

    if (newPacketNumber - self.packetNumber) == 1:
      for i in range(22):
        if i == 2:
          self.speed = self.packet[3] << 8 | self.packet[2]
          self.speed /= 64
        elif self.isDataIndex(i):
          self.distanceIndex = (newPacketNumber - 0xA0) * 4 + i/4 - 1
          if self.distanceIndex >= 0 and self.distanceIndex < 360:
            if (self.packet[i+1] & 0x80) >> 7:
              self.distances[self.distanceIndex] = 0
            else:
              d = self.packet[i] | ((self.packet[i+1] & 0x3F) << 8)

              # cap between limits
              if d < 0 or d > 6000:
                d = 0

              # finally, store valid data
              self.distances[self.distanceIndex] = d

        else:
          #checksum, packet num, and start bit go here
          pass

    self.packetNumber = newPacketNumber

    if self.packetNumber >= 0xF9:
      self.packetNumber = 0x9F

  def isDataIndex(self, index):
    return index == 4 or index == 8 or index == 12 or index == 16

  def checksum(self, data):
    """Compute and return the checksum as an int."""
    # group the data by word, little-endian
    data_list = []
    for t in range(10):
      data_list.append(data[2*t] + (data[2*t+1]<<8))

      # compute the checksum on 32 bits
      chk32 = 0
      for d in data_list:
        chk32 = (chk32 << 1) + d

      # return a value wrapped around on 15bits
      # and truncated to still fit into 15 bits
      # then wrap around to fit into 15 bits
      checksum = (chk32 & 0x7FFF) + (chk32 >> 15)

      # truncate to 15 bits
      checksum = checksum & 0x7FFF

      return int(checksum)

  def getLatestPacket(self):
    return self.packet

  def getLatestDistance(self):
    return self.distances

l = Lidar()

if __name__ == "__main__":
  while True:
    fullSweep = l.read()

File: src/shared/Lidar.cpp
#include <Arduino.h>
#include "Lidar.hpp"

void Lidar::setup() {
    Serial2.begin(115200);
    packetNumber = 0;
    packetIndex = 0;
    distanceIndex = 0;
    startReading = false;
    memset(distances, -1, 360*sizeof(int));
    // Make sure we've tried to read a full sweep
    // before doing anything
    while(!read());
}

bool Lidar::read(){
  if (Serial2.available()) {
    char b = Serial2.read();

    if (startReading) {
      if (packetIndex == 21) {
        startReading = false;
        processEndOfPacket();

        if (distanceIndex >= 357){
          return true;
        }
        return false;
      }
      else {
        packet[packetIndex] = b;
        packetIndex++;
      }
    }

    if (!startReading && (b == (char) 0xfa)){
      if (distanceIndex >= 357){
        memset(distances, -1, 360*sizeof(int));
        /*
        Serial.print("misses: ");
        Serial.print(misses);
        Serial.print(" inv: ");
        Serial.println(invalids);
        */
        misses = 0;
        invalids = 0;
      }
      startReading = true;
      packetIndex = 0;
      packet[packetIndex] = b;
      packetIndex++;
    }
  }
  return false;
}

void Lidar::processEndOfPacket(){
  int newPacketNumber = packet[1];

  if ((newPacketNumber - packetNumber) == 1){
    for (int i=0; i < 22; i++){
      if (isDataIndex(i)){
        distanceIndex = (newPacketNumber - (char)0xA0) * 4 + i/4 - 1;

        if ((distanceIndex >= 0) && (distanceIndex < 360)){
          if ((packet[i+1] & (char)0x80) >> 7){
            distances[distanceIndex] = -1;
            invalids++;
          }
          else {
            int d = (unsigned char)packet[i] | (packet[i+1] << 8);

            if (d <= 140 || d > 6000){
              invalids++;
              d = -1;
            }

            distances[distanceIndex] = d;
          }
        }
      }
    }
  }
  else {
    misses++;
  }

  packetNumber = newPacketNumber;

  if (packetNumber >= (char)0xF9){
    packetNumber = (char)0x9F;
  }
}

bool Lidar::isDataIndex(int index){
  return index == 4 || index == 8 || index == 12 || index == 16;
}

